pulser/sequence.py

```diff
@@ -15,18 +15,71 @@
 from collections import namedtuple
 from collections.abc import Iterable
 import copy
+from functools import wraps
+from itertools import chain
+import json
 import warnings
 
 import numpy as np
 
 import pulser
 from pulser.pulse import Pulse
 from pulser.devices import MockDevice
+from pulser._json_coders import PulserEncoder, PulserDecoder
+from pulser.parametrized import Parametrized, Variable
 from pulser._seq_drawer import draw_sequence
-from pulser.utils import validate_duration
+from pulser.utils import validate_duration, obj_to_dict
 
 # Auxiliary class to store the information in the schedule
 _TimeSlot = namedtuple('_TimeSlot', ['type', 'ti', 'tf', 'targets'])
+# Encodes a sequence building calls
+_Call = namedtuple("_Call", ['name', 'args', 'kwargs'])
+
+
+def _screen(func):
+    """Blocks the call to a function if the Sequence is parametrized."""
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if self.is_parametrized():
+            raise RuntimeError(f"Sequence.{func.__name__} can't be called in"
+                               + " parametrized sequences.")
+        return func(self, *args, **kwargs)
+    return wrapper
+
+
+def _store(func):
+    """Stores any Sequence building call for deferred execution."""
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        def verify_variable(x):
+            if isinstance(x, Parametrized):
+                # If not already, the sequence becomes parametrized
+                self._building = False
+                for name, var in x.variables.items():
+                    if name not in self._variables:
+                        raise ValueError(f"Unknown variable '{name}'.")
+                    elif self._variables[name] is not var:
+                        raise ValueError(
+                            f"{x} has variables that don't come from this "
+                            "Sequence. Use only what's returned by this"
+                            "Sequence's 'declare_variable' method as your"
+                            "variables."
+                            )
+            elif isinstance(x, Iterable) and not isinstance(x, str):
+                # Recursively look for parametrized objs inside the arguments
+                for y in x:
+                    verify_variable(y)
+
+        if self._is_measured and self.is_parametrized():
+            raise SystemError("The sequence has been measured, no further "
+                              "changes are allowed.")
+        # Check if all Parametrized inputs stem from declared variables
+        for x in chain(args, kwargs.values()):
+            verify_variable(x)
+        storage = self._calls if self._building else self._to_build_calls
```

Comment:
I am litlle confused between `self._calls` and `self._to_build_calls`. What is each one used for ? They both seem to be used at the same time. Do we need both ?

Reply:
If it is just about keeping track of which calls need to be build, maybe it would simpler to have an option in the _Call namedtuple ?
I see that when a sequence is parametrized, we use both `self._calls` and `self._to_build_calls` (e.g. with `declare_channel`). Splitting the calls between those two lists might become an issue with teh order of calls at some point no ? Since we don't have a way to keep the order of calls once they are split. I might be missing something though

Reply:
I think you got the main idea, but just to clarify: `self._calls` stores calls that are executed right when they're called, while `self._to_build_calls` are stored to be executed only upon building. 

In general, all calls start out being stored in `_calls` but once the sequence becomes parametrized, the calls are not executed anymore and go instead to `_to_build_calls`. The only exception here is `declare_channel` (`declare_variable` is a different exception since it is never stored), which does not need to be stored in `_to_build_calls` because:
1. It does not accept parametrized parameters.
2. The order at which it occurs doesn't really matter, because the channels are filled independently. 

You'll see that the saving of the `declare_channel` call is not done through the `_store` decorator, but rather "manually". So, in fact, it never ends up in `self._to_build_calls`, always in `_calls`. 

Therefore, the relevant order of the calls, as far as reconstructing an exact replica of the original `Sequence` is concerned, is always well-defined: 
1. Run the `_calls` 
2. Declare the variables 
3. Run the `_to_build_calls`

Reply:
Ok, this works now because declare channels is always done before any parametrized call right ? 
Anyway, I think I got it :+1:

Reply:
Yup, that's it!